articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,119 +38,389 @@ Result result = session.run(queryString);
+All of the above listed APOC procedures include a way to pass a parameter map to the call, providing protection against injection attacks.
+
+== Query Sanitization
+
+Whilst string concatenation for building queries is generally a bad idea, it is not always possible to avoid it.
+Node labels, relationship types and property names are notable examples where parameterization is not supported in Cypher.
+
+In these cases it is important to sanitize user inputs.
+Sanitization is the modification of input to ensure that it is valid, in the case of Cypher, this usually means escaping
+quotation characters or removing delimiter characters which would break out of the context in which they are being used.
+Sanitization should always be done when accepting untrusted external input and may be needed other times, see second order injections for more information.
+
+It is recommended that this sanitization is done at the client level, before passing it down to the database.
+
+=== Escape Characters
+
+Escape characters invoke an alternative meaning on the following characters in the sequence.
+In Cypher, the definition of string literals, regular expressions and naming of types such as node labels can be done
+with the opening and closing of certain characters that can also be used inside the expression given it is escaped correctly.
+
+In all cases, a backslash is escaped using another backslash.
+
+==== String Literals
+
+String literals are started and ended with either a single quote `'` or a double quote `"`.
+These can be escaped using a backslash `\`.
+
+==== Regular Expressions
+
+Regular expressions are started and ended with a single quote `'` which can be escaped using a backslash.
+
+==== Naming
+
+Node labels, relationship types, parameters and variables can be started and ended with a single backtick ```,
+the backtick is escaped using another backtick ```.
+
+For more information on escape characters, see the Cypher Manual on https://neo4j.com/docs/cypher-manual/current/syntax/expressions/[Expressions]
+and https://neo4j.com/docs/cypher-manual/current/syntax/naming/[Naming rules and recommendations].
+
+=== When Sanitization is Necessary
+
+Node labels, relationship types and parameters may contain non-alphabetic characters, including numbers, symbols and whitespace characters, but must be escaped using backticks.
+For example: `node label with spaces`.
+This means that when dynamically building a query using string concatenation, sanitization needs to be done on the escaping of backticks.
+In Cypher, a backtick is escaped using another backtick `&#96;&#96;`.
+
+Here is an example of a simple dynamic label injection attack:
+
+`query = "MATCH (s:School)-[:IN]->(c:&#96;" + cityName + "&#96;) RETURN s;`
+
+With this query we want to search for all schools that are in a certain city, unfortunately our city names are Node Labels,
+so it isn't possible to parameterize the input.
```

Comment:
Perhaps one advice would be to avoid the situation where you need to escape names altogether. And this can probably be achieved by changing the data model, which seems like a very good approach to me.

In this example, instead of having one label per city, you'd have the City label with the schema (name: String) and then users would do `MATCH (c: City { name: $cityName }) RETURN name`. 

I have definitely seen many examples of "bad data modelling" such as this in the query logs of people using apoc.cypher.run to make use of dynamic labels.

I don't think this type of modelling is particularly great. And a little bit of trivia is that the Neo4j dbms only has ~60000 labels names possible which is another reason not to model data this way 🙈 

Let me know what you think.

Reply:
Unfortunately dynamic labels are still something we see from time to time. 

In some cases, this kind of modeling is necessary due to limitations of the product. For example, Max De Marzi's AiportDay graph uses labels for airport + day concatenated, as modeling this way leads to massive performance improvements because filtering on all the relationships of an Airport node by a day property on a relationship isn't performant. If we later get node-centric indexes then there would be no need for such a bizarre model, but until then such design choices remain the only way to get acceptable performance.

Reply:
I think it makes sense to add a comment recommending that the user update their model if possible, in terms of security advice it is good, then the rest of the article still helps in the case where they need it for performance reasons! :)

Reply:
Agreed!